staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go

```diff
@@ -170,17 +170,20 @@ func (c *codec) Encode(obj runtime.Object, w io.Writer) error {
 	case *runtime.Unknown:
 		return c.encoder.Encode(obj, w)
 	case runtime.Unstructured:
-		// avoid conversion roundtrip if GVK is the right one already or is empty (yes, this is a hack, but the old behaviour we rely on in kubectl)
-		objGVK := obj.GetObjectKind().GroupVersionKind()
-		if len(objGVK.Version) == 0 {
-			return c.encoder.Encode(obj, w)
-		}
-		targetGVK, ok := c.encodeVersion.KindForGroupVersionKinds([]schema.GroupVersionKind{objGVK})
-		if !ok {
-			return runtime.NewNotRegisteredGVKErrForTarget(objGVK, c.encodeVersion)
-		}
-		if targetGVK == objGVK {
-			return c.encoder.Encode(obj, w)
+		// Only avoid conversion if unstructed is not a list, giving the converter a chance to convert list items.
+		if !obj.IsList() {
```

Comment:
this is kind of dangerous I fear: the user can have a `list` field in its CRD schema. Those CRs would qualify as lists here, won't they?

Reply:
Do you have a stack trace for when this is needed? Wondering whether we can detect a list earlier.

Reply:
We can actually make sure it is a `UnstructuredList` rather than `obj.IsList()`. I had that variance and it did not change any test result but I guess it is safer. WDYT?

I can retrace my steps to find where this is called.

Reply:
I expect that making the invariant "unstructured things that are lists are always of type UnstructuredList" will make everyone's life better.

Reply:
Changed it to `unstructuredList`. This is the stack trace FYI @sttts:

```
https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Encode at versioning.go#L191
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject at writers.go:97
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated at writers.go:123
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObject at writers.go:52
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.transformResponseObject at response.go:172
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1 at get.go:276
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentHandlerFunc.func1 at metrics.go:238
k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/pkg/apiserver.(*crdHandler).ServeHTTP at customresource_handler.go:262
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP at pathrecorder.go:248
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP at pathrecorder.go:234
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP at handler.go:154
<autogenerated>:2
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1 at authorization.go:64
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1 at maxinflight.go:160
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1 at impersonation.go:49
net/http.HandlerFunc.ServeHTTP at server.go:1947
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1 at timeout.go:97
runtime.goexit at asm_amd64.s:2361
```